[16기 정지원,이혜성][PR_REVIEW_2] : signupview,loginview,utils in user 앱

hyeseong-dev opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한 줄 요약
(여기에 수정 사항에 대한 간략한 한 줄 요약/제목 작성해 주세요.)

> 회원 가입, 로그인 View 작성 완료 

## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)

## 정지원, 이혜성

### 1. PR REVIEW(회원가입, 로그인 - VIEW, URL conf) 
-  병민님 `PR REVIEW` 파악후 재수정하여 깃헙에 다시 푸쉬하기
- 소요시간 : 1시간
- 예상시간 : 15일 20시까지
- 리뷰 사항들 

 - 하나!  \[리뷰 피드백\] (병민멘토님)validation을 위한 함수들은 decorator가 아니기 때문에 decorator directory 밖에 다른 모듈로 옮겨주세요~

  + [완료]  기존 프로젝트 루트 디렉토에서 `decorator/utils.py` 에서`validator`에 대한 부분 전체를  `user/utils.py`파일로 이동시켜 적용함

 - 둘!  \[리뷰 피드백\] 
decorators/utils.py
```python
setattr(request, 'user', User.objects.get(id=['user_id']))
```
line 57에서 db에서 user 레코드를 조회해왔는데, line 58에서 objects.get()를 사용해서 다시 호출하시는 이유가 있나요?

  + [완료]  아래와 같이 수정.
```python
user_id      = User.object.get(id=payload['user_id'])
            setattr(request, 'user', user_id)
```

 - 셋!  \[리뷰 피드백\] 
 -  try문 안으로 넣어주세요~

```python
except jwt.DecodeError:
            return JsonResponse({'MESSAGE':'JWT DECODE ERROR'}, status=400)
        return func(request, *args, **kwargs)
```

  + [완료] 

```python
 except User.DoesNotExist:
            return JsonResponse({"message":"USER_DOES_NOT_EXIST"}, status = 401)
```

 - 넷!!  \[리뷰 피드백\] 
 utils.password_check에서 에러를 핸들링 해주고 있으니, 여기서는 password_check를 호출만하면 되지않나요?
if 문을 지워주세요~
```python
if utils.password_check(password)!=True:
                return utils.password_check(password)
```

  + [완료] 
 함수 호출만함!
```python
utils.password_check(password)
```

 - 다섯!!  \[리뷰 피드백\] 
> token이 어떤 token인지 정확히 명시해주는게 좋습니다.
나중에 refresh_token등 다양한 토큰을 관리해야될 수 있습니다.

```python
   if bcrypt.checkpw(password.encode('utf-8'), login_id.password.encode('utf-8')):
                    user_token = jwt.encode({'user_id': login_id.id}, SECRET_KEY, algorithm=ALGORITHM)
                    return JsonResponse({'TOKEN':user_token}, status=200)
```
  + [완료] 

```python
payload = {"user_id":user.id}
login_token = jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)
return JsonResponse({'MESSAGE':'SUCCESS','TOKEN':login_token}, status=200, safe=True)
```

- 여섯 \[리뷰 피드백\] 
지워주셔도 될 것 같습니다~
```python
signup_db   = User.objects.all()
```
  + [완료] 
삭제 완료


### 2. POST VIEW 및 URL conf 구현
- 예상시간 :16일 23시 59분까지
- 방      법 :
 + 하나. 위스타그램에서 작성한 뷰 및 유알엘 콘프 로직을 참고하여 작성
 + 둘. 동기들이 만든 로직을 참고하여 팀에게 맞는 최선의 로직 구현
 + 셋. 커뮤니티 웹사이트에 글을 남겨서 도움을 얻어본다. 
 + 넷. 모르는 부분이나 애매한 부분은 동기들에게 자문을 얻는다. 직접 DM하여 물어본다. 
 




## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

> 회원 가입 구현 Logic을 더 깔끔하게 작성하고 싶습니다. 
혹시 중복되는 부분이 있을까요?

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [X] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [X] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [X] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [X] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

hyeseong-dev committed:
SignUpView&LoginView 작성완료

hyeseong-dev committed:
UPDATE, JSON MESSAGE INTO UPPER LETTERS

lordmyshepherd commented on decorators/utils.py:
```diff
@@ -0,0 +1,63 @@
+from my_settings import ALGORITHM, SECRET_KEY
+from decorators  import utils
+
+def check_duplication(account, email,phone):
+    return User.objects.filter(Q(account=account) | Q(email=email) | Q(phone=phone)).exists()
+
+def account_validation(account):
+    REGEX_ACCOUNT = '(?i)^(?=.*[a-z])[a-z0-9]{4,20}$'
+    if not re.search(REGEX_ACCOUNT, account):
+        return False
+    return True
+
+def email_validation(email):
+    REGEX_EMAIL = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
+    if not re.search(REGEX_EMAIL, email):
+        return False
+    return True
+
+def username_validation(name):
+    REGEX_KOREAN = ('^[가-힣]')
+    if not re.search(REGEX_KOREAN, name):
+        return False
+    return True
+
+def phone_validation(phone):
+    if len(phone) != 11:
+        return False
+    return True
```
validation을 위한 함수들은 decorator가 아니기 때문에 decorator directory 밖에 다른 모듈로 옮겨주세요~

hyeseong-dev replied:
병민님 감사합니다!😙
다른 모듈로 옮기도록 하겠습니다.

lordmyshepherd commented on decorators/utils.py:
```diff
@@ -0,0 +1,63 @@
+    return True
+
+
+def password_check(password):
+    if len(password)<8 and len(password)>=16:
+        return JsonResponse({'MESSAGE':' PASSWORD SHOULD BE OVER 8 AND UNDER 17 '}, status=412)
+    
+    if not re.findall('[0-9]',password) and (not re.findall('[a-z]',password) or not re.findall('[A-Z]',password)):
+        return JsonResponse({'MESSAGE':'PASSWORD SHOULD BE ONLY ALPHABET AND DIGITS .'}, status=412)
+    
+    if re.search('[@$!%*#?&]+', password) is None:
+        return JsonResponse({'MESSAGE':'PASSWORD INCLUDE AT LEAST 1 CHARACTER'}, status=412)
+    
+    return True
```
마찬가지 입니다.

hyeseong-dev replied:
네에~!!

lordmyshepherd commented on decorators/utils.py:
```diff
@@ -0,0 +1,63 @@
+            token        = request.headers['Authorization']
+            payload      = jwt.decode(token, SECRET_KEY, algorithms=ALGORITHM) 
+            user_id      = User.object.get(id=payload['user_id'])
+            setattr(request, 'user', User.objects.get(id=['user_id']))
```
`line 57`에서 db에서 user 레코드를 조회해왔는데, `line 58`에서 objects.get()를 사용해서 다시 호출하시는 이유가 있나요?
```suggestion
            setattr(request, 'user', User.objects.get(id=['user_id']))
```

hyeseong-dev replied:
중간에 코드 변경하고(위스타 그램에서는  `setattr()` 메서드를 사용(x) ) 리팩토링하지 않아서 생긴 부분이에요.
feedback 주셔서감사합니다.

lordmyshepherd commented on decorators/utils.py:
```diff
@@ -0,0 +1,63 @@
+
+        except jwt.DecodeError:
+            return JsonResponse({'MESSAGE':'JWT DECODE ERROR'}, status=400)
+        return func(request, *args, **kwargs)
```
try문 안으로 넣어주세요~

hyeseong-dev replied:
네! 안으로 바로 넣었습니다. 
그리고  아래부분도 누락하여 추가하였습니다.

```python
  except User.DoesNotExist:
            return JsonResponse({"message":"USER_DOES_NOT_EXIST"}, status = 401)
```

lordmyshepherd commented on user/views.py:
```diff
@@ -1,3 +1,80 @@
+            if not utils.account_validation(account): 
+                return JsonResponse({"MESSAGE":'NOT VALID ACCOUNT'}, status=400)
+
+            if utils.password_check(password)!=True:
+                return utils.password_check(password)
```
`utils.password_check`에서 에러를 핸들링 해주고 있으니, 여기서는 password_check를 호출만하면 되지않나요?
if 문을 지워주세요~

hyeseong-dev replied:
```python
utils.password(password)
```

hyeseong-dev replied:
feedback 주신 부분 반영하여 함수 호출만 하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -1,3 +1,80 @@
+
+                if bcrypt.checkpw(password.encode('utf-8'), login_id.password.encode('utf-8')):
+                    user_token = jwt.encode({'user_id': login_id.id}, SECRET_KEY, algorithm=ALGORITHM)
+                    return JsonResponse({'TOKEN':user_token}, status=200)
```
token이 어떤 token인지 정확히 명시해주는게 좋습니다.
나중에 refresh_token등 다양한 토큰을 관리해야될 수 있습니다.

hyeseong-dev replied:
토큰이름을 변경하도록 하겠습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -1,3 +1,80 @@
+    def post(self, request):
+        try:
+            data        = json.loads(request.body)
+            signup_db   = User.objects.all()
```
지워주셔도 될 것 같습니다~
```suggestion
            signup_db   = User.objects.all()
```

hyeseong-dev replied:
네에~!

hyeseong-dev committed:
[UPDATE][PR REVIEW] SINGN-UP&LOGIN VIEW
1. VALIDATION을 위한 함수들을 별도로 모듈화함
2. 불필요한 코드 리팩토링(예. User.objects.get()
3. 잘못 작성된 login  decorator 수정
4. `if utils.password_check() 문을 사용하지 않고 함수 호출 바로 시행
5. 어떠한 토큰을 사용하였는지 구체화하기 위해 변수명 재설정(login_token)

lordmyshepherd commented:
나중에 유효한 휴대폰 번호 validation을 구현하실 때, 참고해주시면 좋을 것 같아서 코드 남겨놓습니다.

```python
def get_phone_number(phone_number):
    phone_number = (phone_number or '').replace('-', '')
    if phone and len(phone) == 11 and phone[:3] in ['010', '011', '016', '017', '019']:
        return phone
    return ''
```

lordmyshepherd commented:
LGTM 👍
